proposals/NNNN-sequence-grouping.md

```diff
@@ -0,0 +1,250 @@
+| C#            | [`GroupBy`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.groupby?view=net-7.0#system-linq-enumerable-groupby) | [`ToDictionary`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.todictionary?view=net-7.0#system-linq-enumerable-todictionary) |
+| Rust          | [`group_by`](https://doc.rust-lang.org/std/primitive.slice.html#method.group_by) | - |
+| Ruby          | [`group_by`](https://ruby-doc.org/3.2.2/Enumerable.html#method-i-group_by) | [`index_by`](https://rubydoc.info/gems/activesupport/7.0.5/Enumerable#index_by-instance_method) |
+| Python        | [`groupby`](https://docs.python.org/3/library/itertools.html#itertools.groupby) | [dict comprehensions](https://peps.python.org/pep-0274/) |
+| PHP (Laravel) | [`groupBy`](https://laravel.com/docs/10.x/collections#method-groupby) | [`keyBy`](https://laravel.com/docs/10.x/collections#method-keyby) |
+
+## Proposed solution
+
+This proposal introduces 2 new methods on `Sequence`. Here are simple examples of their usages:
+
+```swift
+let digitsGroupedByMod3 = (0...9).grouped(by: { $0 % 3 })
+// Results in:
+[
+	0: [0, 3, 6, 9],
+	1: [1, 4, 7],
+	2: [2, 5, 8],
+]
+
+let fruitsByFirstLetter = ["Apple", "Banana", "Cherry"].keyed(by: { $0.first! })
+// Results in:
+[
+	"A": "Apple",
+	"B": "Banana",
+	"C": "Cherry",
+]
+```
+
+## Detailed design
+
+```swift
+extension Sequence {
+	/// Groups up elements of `self` into a new Dictionary,
+	/// whose values are Arrays of grouped elements,
+	/// each keyed by the group key returned by the given closure.
+	/// - Parameters:
+	///   - keyForValue: A closure that returns a key for each element in
+	///     `self`.
+	/// - Returns: A dictionary containing grouped elements of self, keyed by
+	///     the keys derived by the `keyForValue` closure.
+	func grouped<GroupKey>(
+		by keyForValue: (Element) throws -> GroupKey
+	) rethrows -> [GroupKey: [Element]]
+	
+	/// Creates a new Dictionary from the elements of `self`, keyed by the
+	/// results returned by the given `keyForValue` closure. As the dictionary is
+	/// built, the initializer calls the `combine` closure with the current and
+	/// new values for any duplicate keys. Pass a closure as `combine` that
+	/// returns the value to use in the resulting dictionary: The closure can
+	/// choose between the two values, combine them to produce a new value, or
+	/// even throw an error.
+	///
+	/// If no `combine` closure is provided, deriving the same duplicate key for
+	/// more than one element of self results in a runtime error.
+	///
+	/// - Parameters:
+	///   - keyForValue: A closure that returns a key for each element in
+	///     `self`.
+	///     dictionary.
+	///   - combine: A closure that is called with the values for any duplicate
+	///     keys that are encountered. The closure returns the desired value for
+	///     the final dictionary.
+    func keyed<Key>(
+		by keyForValue: (Element) -> Key,
```

Comment:
Why is this not also `throws`?

Reply:
Nice catch. It is in [the implementation](https://github.com/apple/swift-algorithms/pull/197/files#diff-a520582aa3661c878867008f0cfe2b90d7c2e9c0bd2908a27b7203ba69ee2dc4R46). I've updated this :)